### PR TITLE
issue/1523: added xapi model state functions

### PR DESCRIPTION
**Pull request by oliverfoster**
as per #1523

### CODE REVIEW

File: src/core/js/models/componentModel.js [hunk 1]
```diff
@@ -5,7 +5,45 @@ define([
...
+            state = _.pick.apply(_, args);
+
+            this.set(state);
+
+            return this;
+            
+        },
+
+        triggerState: function() {
+            Adapt.tirgger("state:change", this.getState());
```
Comment (moloko) [inline src/core/js/models/componentModel.js]: typo `tirgger`...

File: src/core/js/libraries/underscore.results.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            /*
+                if the resolved result is empty, replace with the same type as the withData
+            */
+            if (!result) {
+                switch (withType) {
+                    case "array":
+                        result = [];
+                        break;
+                    case "object":
+                        retuls = {};
```
Comment (brian-learningpool) [inline src/core/js/libraries/underscore.results.js]: Typo.

File: src/core/js/libraries/underscore.results.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            */
+            var defaultType = (defaultResult instanceof Array ? "array" : typeof defaultResult);
+
+            if (!result) {
+                switch (defaultType) {
+                    case "array":
+                        result = [];
+                        break;
+                    case "object":
+                        retuls = {};
```
Comment (brian-learningpool) [inline src/core/js/libraries/underscore.results.js]: Typo?

Comment (oliverfoster) [inline src/core/js/libraries/underscore.results.js]: Yup

File: src/core/js/models/adaptModel.js [hunk 4]
```diff
@@ -52,7 +58,54 @@ define([
...
+                }, this),
+                _.debounce(_.bind(function() {
+                    
+                    // Trigger original function
+                    originalTrackableStateFunction.apply(this);
+
+                    // Unset waiting flag
+                    this.triggerTrackableState.isQueued = false;
+
+                }, this), 17)
```
Comment (brian-learningpool) [inline src/core/js/models/adaptModel.js]: The 17 here is a magic number.  Also, why 17?

Comment (oliverfoster) [reply src/core/js/models/adaptModel.js]: 60fps ish. anywhere between 17ms and 40ms is usually good for stuff like this (25-60fps).
a defer would probably suffice but i wanted to group as many model changes as i could into one trigger, these number allow for a few short asynchronous model changes to occur before a callback without it seeming too laggy. > 250ms is probably too laggy.

File: src/core/js/models/adaptModel.js [hunk 5]
```diff
@@ -52,7 +58,54 @@ define([
...
+                // Skip if trigger queued or adapt hasn't started yet
+                if (this.triggerTrackableState.isQueued || !Adapt.attributes._isStarted) {
+                    return
+                }
+
+                // Check that property is trackable
+                var trackable = _.result(this, 'trackable', []);
+                if (!_.keys(model.changed).find(function(item, index) {
+                    return _.contains(trackable, item);
+                })) return;
```
Comment (tomgreenfield) [inline src/core/js/models/adaptModel.js]: This is quite difficult to read. Could we change to something like:
```js
var isTrackable = _.keys(model.changed).find(function(item, index) {
     return _.contains(_.result(this, 'trackable', []), item);
}.bind(this));

if (isTrackable) {
    this.triggerTrackableState();
}
```
❓

File: src/core/js/models/adaptModel.js [hunk 6]
```diff
@@ -52,7 +58,54 @@ define([
...
+
+                }, this), 17)
+            );
+
+            // Listen to model changes, trigger trackable state change when appropriate
+            this.listenTo(this, "change", function(model, value) {
+
+                // Skip if trigger queued or adapt hasn't started yet
+                if (this.triggerTrackableState.isQueued || !Adapt.attributes._isStarted) {
+                    return
```
Comment (tomgreenfield) [inline src/core/js/models/adaptModel.js]: Missing semi-colon.

**Pull request merged**